// antd-editor/src/components/InsertLinkModal.js
import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Form, Select, Input } from 'antd';
class InsertLinkModal extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    const formItemLayout = {
      labelCol: { span: 4 },
      wrapperCol: { span: 20 }
    };
    const { getFieldDecorator } = this.props.form;
    return (
      <Modal
        title="插入链接"
        visible={this.props.visible}
        onCancel={this.props.onCancel}
        maskClosable={false}
        onOk={() => {
          this.props.form.validateFieldsAndScroll((err, values) => {
            if (!err) {
              this.props.onConfirm(values);
              this.props.form.resetFields();
            }
          })
        }}
      >
        <Form>
          <Form.Item
            label="链接地址"
            {...formItemLayout}
            hasFeedback
          >
            {getFieldDecorator('href', {
              rules: [
                {
                  required: true,
                  message: '链接地址不能为空'
                }
              ]
            })(
              <Input placeholder="请输入链接地址"/>
            )}
          </Form.Item>
          {
            !this.props.edit ?
              <Form.Item
                label="链接文字"
                {...formItemLayout}
                hasFeedback
              >
                {getFieldDecorator('text', {
                  rules: [
                    {
                      required: !this.props.edit,
                      message: '链接文字不能为空'
                    }
                  ]
                })(
                  <Input placeholder="请输入链接文字"/>
                )}
              </Form.Item>
              : null
          }
          <Form.Item
            label="打开方式"
            {...formItemLayout}
            hasFeedback
          >
            {getFieldDecorator('target', {})(
              <Select placeholder="请选择打开方式" allowClear={true}>
                <Select.Option value="_blank">在新窗口打开</Select.Option>
              </Select>
            )}
          </Form.Item>
        </Form>
      </Modal>
    );
  }
}
InsertLinkModal.propTypes = {
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  edit: PropTypes.bool
};
InsertLinkModal.defaultProps = {
  edit: true
};
export default Form.create()(InsertLinkModal);
