// antd-editor/lib/components/InsertImageModal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _style9 = require('antd/lib/modal/style');

var _modal = require('antd/lib/modal');

var _modal2 = _interopRequireDefault(_modal);

var _style10 = require('antd/lib/row/style');

var _row = require('antd/lib/row');

var _row2 = _interopRequireDefault(_row);

var _style11 = require('antd/lib/button/style');

var _button = require('antd/lib/button');

var _button2 = _interopRequireDefault(_button);

var _style12 = require('antd/lib/col/style');

var _col = require('antd/lib/col');

var _col2 = _interopRequireDefault(_col);

var _style13 = require('antd/lib/form/style');

var _form = require('antd/lib/form');

var _form2 = _interopRequireDefault(_form);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _style14 = require('antd/lib/input/style');

var _input = require('antd/lib/input');

var _input2 = _interopRequireDefault(_input);

var _style15 = require('antd/lib/upload/style');

var _upload = require('antd/lib/upload');

var _upload2 = _interopRequireDefault(_upload);

var _style16 = require('antd/lib/icon/style');

var _icon = require('antd/lib/icon');

var _icon2 = _interopRequireDefault(_icon);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var InsertImageModal = function (_React$Component) {
  (0, _inherits3.default)(InsertImageModal, _React$Component);

  function InsertImageModal(props) {
    (0, _classCallCheck3.default)(this, InsertImageModal);

    var _this = (0, _possibleConstructorReturn3.default)(this, (InsertImageModal.__proto__ || (0, _getPrototypeOf2.default)(InsertImageModal)).call(this, props));

    _this.state = {
      image: null
    };
    return _this;
  }

  (0, _createClass3.default)(InsertImageModal, [{
    key: 'setImage',
    value: function setImage(url) {
      this.setState({ image: url });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var formItemLayout = {
        labelCol: { span: 4 },
        wrapperCol: { span: 20 }
      };
      var getFieldDecorator = this.props.form.getFieldDecorator;

      return _react2.default.createElement(
        _modal2.default,
        {
          title: '\u63D2\u5165\u56FE\u7247',
          visible: this.props.visible,
          maskClosable: false,
          onCancel: this.props.onCancel,
          onOk: function onOk() {
            _this2.props.form.validateFieldsAndScroll(function (err, values) {
              if (!err) {
                _this2.props.onConfirm(values);
                _this2.props.form.resetFields();
                _this2.setState({ image: null });
              }
            });
          }
        },
        _react2.default.createElement(
          _upload2.default.Dragger,
          {
            beforeUpload: function beforeUpload(file, fileList) {
              var callback = function callback(url) {
                _this2.props.form.setFields({ 'src': { value: url } });
                _this2.setImage(url);
              };
              _this2.props.onUpload(file, fileList, callback);
              return false;
            }
          },
          this.state.image || this.props.currentImage ? _react2.default.createElement('img', { src: this.state.image || this.props.currentImage.src,
            style: { display: 'block', minHeight: 122, maxHeight: 400, maxWidth: '100%', margin: '0 auto' } }) : _react2.default.createElement(
            'div',
            { style: { margin: 16 } },
            _react2.default.createElement(
              'p',
              { style: { fontSize: 48, textAlign: 'center', color: '#108ee9' } },
              _react2.default.createElement(_icon2.default, { type: 'cloud-upload-o' })
            ),
            _react2.default.createElement(
              'p',
              { style: { color: '#999' } },
              '\u70B9\u51FB\u6216\u8005\u62D6\u52A8\u56FE\u7247\u81F3\u6B64\u4E0A\u4F20'
            )
          )
        ),
        _react2.default.createElement(
          _form2.default,
          { style: { marginTop: 16 } },
          getFieldDecorator('key', {
            initialValue: this.props.currentImage && this.props.currentImage.key
          })(_react2.default.createElement(_input2.default, { type: 'hidden' })),
          _react2.default.createElement(
            _form2.default.Item,
            (0, _extends3.default)({
              label: '\u56FE\u7247\u5730\u5740'
            }, formItemLayout, {
              hasFeedback: true
            }),
            getFieldDecorator('src', {
              rules: [{
                required: true,
                message: '图片地址不能为空'
              }],
              initialValue: this.props.currentImage && this.props.currentImage.src
            })(_react2.default.createElement(_input2.default, { placeholder: '\u8BF7\u8F93\u5165\u56FE\u7247\u5730\u5740', onBlur: function onBlur(e) {
                _this2.setImage(e.target.value);
              } }))
          ),
          _react2.default.createElement(
            _form2.default.Item,
            (0, _extends3.default)({
              label: '\u66FF\u4EE3\u6587\u5B57'
            }, formItemLayout, {
              hasFeedback: true
            }),
            getFieldDecorator('alt', {
              initialValue: this.props.currentImage && this.props.currentImage.alt
            })(_react2.default.createElement(_input2.default, { placeholder: '\u8BF7\u8F93\u5165\u66FF\u4EE3\u6587\u5B57' }))
          ),
          _react2.default.createElement(
            _row2.default,
            null,
            _react2.default.createElement(
              _col2.default,
              { span: 8 },
              _react2.default.createElement(
                _form2.default.Item,
                {
                  label: '\u56FE\u7247\u5BBD\u5EA6',
                  hasFeedback: true,
                  labelCol: { span: 12 },
                  wrapperCol: { span: 12 }
                },
                getFieldDecorator('width', {
                  initialValue: this.props.currentImage && this.props.currentImage.width
                })(_react2.default.createElement(_input2.default, { size: 'large', placeholder: '\u8BF7\u8F93\u5165\u5BBD\u5EA6' }))
              )
            ),
            _react2.default.createElement(
              _col2.default,
              { span: 8 },
              _react2.default.createElement(
                _form2.default.Item,
                {
                  label: '\u56FE\u7247\u9AD8\u5EA6',
                  hasFeedback: true,
                  labelCol: { span: 12 },
                  wrapperCol: { span: 12 }
                },
                getFieldDecorator('height', {
                  initialValue: this.props.currentImage && this.props.currentImage.height
                })(_react2.default.createElement(_input2.default, { size: 'large', placeholder: '\u8BF7\u8F93\u5165\u9AD8\u5EA6' }))
              )
            ),
            _react2.default.createElement(
              _col2.default,
              { span: 8 },
              _react2.default.createElement(
                _button2.default,
                { size: 'large', style: { marginLeft: 24 }, onClick: function onClick() {
                    _this2.props.form.setFields({ 'width': { value: '100%' }, 'height': { value: undefined } });
                  } },
                '\u81EA\u9002\u5E94\u5BBD\u9AD8'
              )
            )
          )
        )
      );
    }
  }]);
  return InsertImageModal;
}(_react2.default.Component);

InsertImageModal.propTypes = {
  visible: _propTypes2.default.bool.isRequired,
  onCancel: _propTypes2.default.func.isRequired,
  onConfirm: _propTypes2.default.func.isRequired,
  currentImage: _propTypes2.default.object,
  onUpload: _propTypes2.default.func
};
exports.default = _form2.default.create()(InsertImageModal);
