// antd-editor/lib/components/InsertLinkModal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _style5 = require('antd/lib/modal/style');

var _modal = require('antd/lib/modal');

var _modal2 = _interopRequireDefault(_modal);

var _style6 = require('antd/lib/select/style');

var _select = require('antd/lib/select');

var _select2 = _interopRequireDefault(_select);

var _style7 = require('antd/lib/form/style');

var _form = require('antd/lib/form');

var _form2 = _interopRequireDefault(_form);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _style8 = require('antd/lib/input/style');

var _input = require('antd/lib/input');

var _input2 = _interopRequireDefault(_input);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var InsertLinkModal = function (_React$Component) {
  (0, _inherits3.default)(InsertLinkModal, _React$Component);

  function InsertLinkModal(props) {
    (0, _classCallCheck3.default)(this, InsertLinkModal);
    return (0, _possibleConstructorReturn3.default)(this, (InsertLinkModal.__proto__ || (0, _getPrototypeOf2.default)(InsertLinkModal)).call(this, props));
  }

  (0, _createClass3.default)(InsertLinkModal, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var formItemLayout = {
        labelCol: { span: 4 },
        wrapperCol: { span: 20 }
      };
      var getFieldDecorator = this.props.form.getFieldDecorator;

      return _react2.default.createElement(
        _modal2.default,
        {
          title: '\u63D2\u5165\u94FE\u63A5',
          visible: this.props.visible,
          onCancel: this.props.onCancel,
          maskClosable: false,
          onOk: function onOk() {
            _this2.props.form.validateFieldsAndScroll(function (err, values) {
              if (!err) {
                _this2.props.onConfirm(values);
                _this2.props.form.resetFields();
              }
            });
          }
        },
        _react2.default.createElement(
          _form2.default,
          null,
          _react2.default.createElement(
            _form2.default.Item,
            (0, _extends3.default)({
              label: '\u94FE\u63A5\u5730\u5740'
            }, formItemLayout, {
              hasFeedback: true
            }),
            getFieldDecorator('href', {
              rules: [{
                required: true,
                message: '链接地址不能为空'
              }]
            })(_react2.default.createElement(_input2.default, { placeholder: '\u8BF7\u8F93\u5165\u94FE\u63A5\u5730\u5740' }))
          ),
          !this.props.edit ? _react2.default.createElement(
            _form2.default.Item,
            (0, _extends3.default)({
              label: '\u94FE\u63A5\u6587\u5B57'
            }, formItemLayout, {
              hasFeedback: true
            }),
            getFieldDecorator('text', {
              rules: [{
                required: !this.props.edit,
                message: '链接文字不能为空'
              }]
            })(_react2.default.createElement(_input2.default, { placeholder: '\u8BF7\u8F93\u5165\u94FE\u63A5\u6587\u5B57' }))
          ) : null,
          _react2.default.createElement(
            _form2.default.Item,
            (0, _extends3.default)({
              label: '\u6253\u5F00\u65B9\u5F0F'
            }, formItemLayout, {
              hasFeedback: true
            }),
            getFieldDecorator('target', {})(_react2.default.createElement(
              _select2.default,
              { placeholder: '\u8BF7\u9009\u62E9\u6253\u5F00\u65B9\u5F0F', allowClear: true },
              _react2.default.createElement(
                _select2.default.Option,
                { value: '_blank' },
                '\u5728\u65B0\u7A97\u53E3\u6253\u5F00'
              )
            ))
          )
        )
      );
    }
  }]);
  return InsertLinkModal;
}(_react2.default.Component);

InsertLinkModal.propTypes = {
  visible: _propTypes2.default.bool.isRequired,
  onCancel: _propTypes2.default.func.isRequired,
  onConfirm: _propTypes2.default.func.isRequired,
  edit: _propTypes2.default.bool
};
InsertLinkModal.defaultProps = {
  edit: true
};
exports.default = _form2.default.create()(InsertLinkModal);
